File: execcom_web/src/core/permission-engine.ts
import { AppRole, appRoleFromString, ExecomFeature, BudgetAuthorityPosts } from './constants';
import { UserModel, ExecomMemberModel, ExecomPermissionModel } from '../models/types';

export class PermissionEngine {
  readonly user: UserModel;
  readonly userExecomMemberships: ExecomMemberModel[];
  readonly execomPermissions: Record<number, ExecomPermissionModel[]>;

  constructor(
    user: UserModel,
    userExecomMemberships: ExecomMemberModel[] = [],
    execomPermissions: Record<number, ExecomPermissionModel[]> = {}
  ) {
    this.user = user;
    this.userExecomMemberships = userExecomMemberships;
    this.execomPermissions = execomPermissions;
  }

  get role(): AppRole {
    return appRoleFromString(this.user.role);
  }

  // Tier getters
  get isTier1(): boolean { return this.role === AppRole.chairman || this.role === AppRole.viceChairman; }
  get isTier2(): boolean { return this.role === AppRole.coreExeccom || this.role === AppRole.facultyAdvisor; }
  get isTier3(): boolean { return this.role === AppRole.forumExeccom; }
  get isTier4(): boolean { return this.role === AppRole.panel; }
  get isTier5(): boolean { return this.role === AppRole.restricted; }
  get isTier6(): boolean { return this.role === AppRole.member; }

  get isAtLeastTier1(): boolean { return this.role >= AppRole.viceChairman; }
  get isAtLeastTier2(): boolean { return this.role >= AppRole.coreExeccom; }
  get isAtLeastTier3(): boolean { return this.role >= AppRole.forumExeccom; }
  get isAtLeastTier4(): boolean { return this.role >= AppRole.panel; }

  get isEffectivelyTier1(): boolean {
    return this.isTier1 || (this.isTier2 && this.user.is_sudo);
  }

  // Actions
  get canAddMembers(): boolean { return this._isCommitteeLead; }
  get canRemoveMembers(): boolean { return this.isAtLeastTier1; }
  get canEditMembers(): boolean { return this.isTier1; }
  get canAssignRoles(): boolean { return this.isAtLeastTier1; }
  get canManageExecom(): boolean { return this.isAtLeastTier1; }
  get canManageGlobalPermissions(): boolean { return this.role === AppRole.chairman; }
  get canManageExecomPermissions(): boolean { return this.isAtLeastTier1; }
  get canManagePermissions(): boolean { return this.canManageGlobalPermissions || this.canManageExecomPermissions; }

  get _isCommitteeLead(): boolean {
    if (this.isAtLeastTier1) return true;
    if (this.isAtLeastTier2) return true;
    if (this.isTier3) {
      const p = (this.user.post || '').toLowerCase();
      return p.includes('chair') || p.includes('secretary');
    }
    return false;
  }

  get _isTop4(): boolean {
    if (this.isTier1) return true;
    if (this.isTier2) {
      const p = (this.user.post || '').toLowerCase();
      return p === 'secretary' || p === 'treasurer';
    }
    return false;
  }

  canManageMembersInExecom(execomId: number): boolean {
    if (this.isAtLeastTier1) return true;
    if (this.isAtLeastTier2 && this.isMemberOfExecom(execomId)) return true;
    const fRole = (this.execomRoleIn(execomId) || '').toLowerCase();
    return fRole.includes('chair') || fRole.includes('head');
  }

  isFeatureEnabledGlobally(feature: string): boolean {
    const perms = this.execomPermissions[0] || [];
    const perm = perms.find((p) => p.feature === feature);
    return perm ? perm.allowed : false;
  }

  // Feature limits
  get canCreateEvents(): boolean { return this.isAtLeastTier3; }
  get canReadReports(): boolean { return this.isAtLeastTier2; }
  get canUploadReports(): boolean { return this.isAtLeastTier4; }
  get canViewTotalBudget(): boolean { return this.isAtLeastTier2; }
  get canAccessScopedBudget(): boolean { return this.isAtLeastTier3; }
  get canViewMembers(): boolean { return this.role !== AppRole.restricted; }
  get isPanel(): boolean { return this.isTier4; }
  get canRequestBudget(): boolean { return this.isAtLeastTier3; }
  get canManageBudget(): boolean { return this._isTop4; }
  get canViewInternalAnnouncements(): boolean { return this.isAtLeastTier3; }
  get canManageAnnouncements(): boolean { return this.isAtLeastTier2; }
  get canOverride(): boolean { return this.role === AppRole.chairman; }
  get canAccessChat(): boolean { return this.isAtLeastTier4; }

  get canApproveBudget(): boolean {
    if (this.isEffectivelyTier1) return true;
    if (this.isTier2) return BudgetAuthorityPosts.hasBudgetAuthority(this.user.post);
    return false;
  }

  // Execom Scopes
  isMemberOfExecom(execomId: number): boolean {
    return this.userExecomMemberships.some((m) => m.execom_id === execomId);
  }

  execomRoleIn(execomId: number): string | undefined {
    const membership = this.userExecomMemberships.find((m) => m.execom_id === execomId);
    return membership?.execom_role;
  }

  canDoInExecom(execomId: number, feature: string): boolean {
    if (this.isEffectivelyTier1) return true;

    const perms = this.execomPermissions[execomId] || [];
    const perm = perms.find((p) => p.feature === feature);

    if (this.isAtLeastTier2) {
      if (perm !== undefined) return perm.allowed;
      return true;
    }

    if (!this.isMemberOfExecom(execomId)) return false;

    if (this.isAtLeastTier3) {
      if (perm !== undefined) return perm.allowed;
      return true;
    }

    return perm ? perm.allowed : false;
  }

  canCreateEventInExecom(execomId: number): boolean {
    return this.canCreateEvents && this.canDoInExecom(execomId, ExecomFeature.createEvents);
  }

  canUploadReportInExecom(execomId: number): boolean {
    return this.canUploadReports && this.canDoInExecom(execomId, ExecomFeature.uploadReports);
  }

  canViewBudgetInExecom(execomId: number): boolean {
    return this.canAccessScopedBudget && this.canDoInExecom(execomId, ExecomFeature.viewBudget);
  }

  get userExecomIds(): number[] {
    return this.userExecomMemberships.map((m) => m.execom_id);
  }
}

File: execcom_app/lib/core/permission_engine.dart
import '../models/user_model.dart';
import '../models/folder_model.dart';
import 'constants.dart';

/// Central permission resolver.
/// Checks: role level → folder membership → folder permissions → post-specific rules.
class PermissionEngine {
  final UserModel user;
  final List<FolderMemberModel> userFolderMemberships;
  final Map<int, List<FolderPermissionModel>> folderPermissions;

  const PermissionEngine({
    required this.user,
    this.userFolderMemberships = const [],
    this.folderPermissions = const {},
  });

  AppRole get role => AppRole.fromString(user.role);

  // ── Tier-based getters ──

  bool get isTier1 => role == AppRole.chairman || role == AppRole.viceChairman;
  bool get isTier2 => role == AppRole.coreExeccom;
  bool get isTier3 => role == AppRole.forumExeccom;
  bool get isTier4 => role == AppRole.panel;
  bool get isTier5 => role == AppRole.restricted;
  bool get isTier6 => role == AppRole.member;

  bool get isAtLeastTier1 => role >= AppRole.viceChairman;
  bool get isAtLeastTier2 => role >= AppRole.coreExeccom;
  bool get isAtLeastTier3 => role >= AppRole.forumExeccom;
  bool get isAtLeastTier4 => role >= AppRole.panel;

  /// Effective Tier 1 (includes Tier 2 with Chairman's Sudo grant)
  bool get isEffectivelyTier1 => isTier1 || (isTier2 && user.isSudo);

  // ── Action permissions ──

  bool get canAddMembers => _isCommitteeLead;
  bool get canRemoveMembers => isAtLeastTier1;
  bool get canEditMembers => isTier1;
  bool get canAssignRoles => isAtLeastTier1;
  bool get canManageFolders => isAtLeastTier1;
  bool get canManageGlobalPermissions => role == AppRole.chairman;
  bool get canManageFolderPermissions => isAtLeastTier1;
  bool get canManagePermissions => canManageGlobalPermissions || canManageFolderPermissions;

  /// Fundamental check for committee leadership (Heads, Chairs, Secretaries)
  bool get _isCommitteeLead {
    if (isAtLeastTier1) return true;
    if (isAtLeastTier2) return true;
    if (isTier3) {
      final p = user.post?.toLowerCase() ?? '';
      return p.contains('chair') || p.contains('secretary');
    }
    return false;
  }

  /// Top 4 roles with full administrative and budget authority:
  /// Chairman, Vice Chairman, Secretary, Treasurer.
  bool get _isTop4 {
    if (isTier1) return true;
    if (isTier2) {
      final p = user.post?.toLowerCase() ?? '';
      return p == 'secretary' || p == 'treasurer';
    }
    return false;
  }

  /// Can manage members within a specific folder/forum
  bool canManageMembersInFolder(int folderId) {
    if (isAtLeastTier1) return true;
    // Core members assigned to a folder can manage its members
    if (isAtLeastTier2 && isMemberOfFolder(folderId)) return true;
    // Forum Chairs/Heads can manage their own members
    final fRole = folderRoleIn(folderId)?.toLowerCase() ?? '';
    return fRole.contains('chair') || fRole.contains('head');
  }

  /// Is a specific feature allowed globally (folder_id: 0)
  bool isFeatureEnabledGlobally(String feature) {
    final perms = folderPermissions[0] ?? [];
    final perm = perms.where((p) => p.feature == feature).firstOrNull;
    return perm?.allowed ?? false;
  }

  // ── Mandated Permissions ──

  /// 1. Event creation privileges restricted exclusively to Tier 3 and above.
  bool get canCreateEvents => isAtLeastTier3;

  /// 2. Report viewing access is limited to Tier 2 and all tiers above.
  bool get canReadReports => isAtLeastTier2;

  bool get canUploadReports => isAtLeastTier4;

  /// 5. Full organizational budget viewing access is restricted to Tier 2 and above.
  bool get canViewTotalBudget => isAtLeastTier2;

  /// 6. Tier 3 (Forum-Execom) may view their forum budget. No activation gate.
  bool get canAccessScopedBudget => isAtLeastTier3;

  /// Anyone EXCEPT restricted members can view the org member list (basic details only).
  bool get canViewMembers => role != AppRole.restricted;

  // ── Legacy/Other Access ──
  bool get isPanel => isTier4;
  bool get canRequestBudget => isAtLeastTier3;
  bool get canManageBudget => _isTop4;

  bool get canViewInternalAnnouncements => isAtLeastTier3;
  bool get canManageAnnouncements => isAtLeastTier2;
  bool get canOverride => role == AppRole.chairman;
  bool get canAccessChat => isAtLeastTier4;

  /// Budget approve/reject: only effectively Tier 1
  /// or Tier 2 with specific posts (Treasurer, Sub-Treasurer)
  bool get canApproveBudget {
    if (isEffectivelyTier1) return true;
    if (isTier2) return BudgetAuthorityPosts.hasBudgetAuthority(user.post);
    return false;
  }

  // ── Folder-scoped permissions ──

  /// Is this user a member of a specific folder?
  bool isMemberOfFolder(int folderId) =>
      userFolderMemberships.any((m) => m.folderId == folderId);

  /// Get the user's role within a folder
  String? folderRoleIn(int folderId) {
    final membership = userFolderMemberships
        .where((m) => m.folderId == folderId)
        .firstOrNull;
    return membership?.folderRole;
  }

  /// Check if a folder feature is allowed for this user
  bool canDoInFolder(int folderId, String feature) {
    // Effective Tier 1 can do everything in any folder
    if (isEffectivelyTier1) return true;

    // Check folder-specific permission toggle
    final perms = folderPermissions[folderId] ?? [];
    final perm = perms.where((p) => p.feature == feature).firstOrNull;
    
    // If not a member, they can only see if it's publicly allowed for their role
    // Core members and above default to true everywhere if not explicitly restricted
    if (isAtLeastTier2) {
      if (perm != null) return perm.allowed;
      return true; 
    }

    if (!isMemberOfFolder(folderId)) return false;

    // Forum Execcom (Tier 3) default to true in their own folder if not explicitly restricted
    if (isAtLeastTier3) {
      if (perm != null) return perm.allowed;
      return true;
    }

    return perm?.allowed ?? false;
  }

  /// Can create events in a specific folder
  bool canCreateEventInFolder(int folderId) =>
      canCreateEvents && canDoInFolder(folderId, FolderFeature.createEvents);

  /// Can upload reports in a specific folder
  bool canUploadReportInFolder(int folderId) =>
      canUploadReports && canDoInFolder(folderId, FolderFeature.uploadReports);

  /// Can view budget in a specific folder
  bool canViewBudgetInFolder(int folderId) =>
      canAccessScopedBudget && canDoInFolder(folderId, FolderFeature.viewBudget);

  /// Get list of folder IDs this user belongs to
  List<int> get userFolderIds =>
      userFolderMemberships.map((m) => m.folderId).toList();
}
